<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Libraries-and-Datasets" data-toc-modified-id="Libraries-and-Datasets-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Libraries and Datasets</a></span></li><li><span><a href="#Looking-at-engine_faults" data-toc-modified-id="Looking-at-engine_faults-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Looking at <code>engine_faults</code></a></span><ul class="toc-item"><li><span><a href="#Provided-Data-Dictionary-for-engine_faults" data-toc-modified-id="Provided-Data-Dictionary-for-engine_faults-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Provided Data Dictionary for <code>engine_faults</code></a></span></li><li><span><a href="#Data-Cleanups-engine_Faults" data-toc-modified-id="Data-Cleanups-engine_Faults-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Data Cleanups <code>engine_Faults</code></a></span><ul class="toc-item"><li><span><a href="#Remove-faults-where-equipmentID-has-more-than-5-characters" data-toc-modified-id="Remove-faults-where-equipmentID-has-more-than-5-characters-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Remove faults where equipmentID has more than 5 characters</a></span></li><li><span><a href="#Remove-faults-in-the-vicinity-of-service-location" data-toc-modified-id="Remove-faults-in-the-vicinity-of-service-location-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Remove faults in the vicinity of service location</a></span></li><li><span><a href="#Remove-faults-where-active-is-set-to-False" data-toc-modified-id="Remove-faults-where-active-is-set-to-False-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Remove faults where <code>active</code> is set to <code>False</code></a></span></li></ul></li><li><span><a href="#Final-engine_faults-before-merge" data-toc-modified-id="Final-engine_faults-before-merge-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Final <code>engine_faults</code> before merge</a></span></li></ul></li><li><span><a href="#Looking-at-diagnostics" data-toc-modified-id="Looking-at-diagnostics-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Looking at <code>diagnostics</code></a></span><ul class="toc-item"><li><span><a href="#Data-Cleanups:-diagnostics" data-toc-modified-id="Data-Cleanups:-diagnostics-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Data Cleanups: <code>diagnostics</code></a></span></li></ul></li><li><span><a href="#Looking-At-service_fault_codes" data-toc-modified-id="Looking-At-service_fault_codes-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Looking At <code>service_fault_codes</code></a></span><ul class="toc-item"><li><span><a href="#Data-Cleanups-service_fault_codes" data-toc-modified-id="Data-Cleanups-service_fault_codes-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Data Cleanups <code>service_fault_codes</code></a></span></li></ul></li></ul></div>

## Libraries and Datasets

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
engine_faults = pd.read_csv(
    "../data/J1939Faults.csv", 
    low_memory=False
)

In [3]:
diagnostics = pd.read_csv(
    "../data/VehicleDiagnosticOnboardData.csv", 
    low_memory=False
)

In [4]:
service_fault_codes = pd.read_excel(
    "../data/Service Fault Codes_1_0_0_167.xlsx", 
    sheet_name="Published 14602 App A core II",
    engine="openpyxl"
)

C:\Users\maeva\anaconda3\lib\site-packages\openpyxl\worksheet\_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed
  warn(msg)


In [5]:
# This is old data: We can't use it
# vehicle_makes = pd.read_excel(
#     '../data/Vehicle_Make.xlsx', 
#     sheet_name='Sheet1',
#     engine="openpyxl"
# )

## Looking at `engine_faults`

In [6]:
display(engine_faults.shape)
display(engine_faults.head())
display(engine_faults.tail())
display(engine_faults.info())

(1187335, 20)

RecordID  ESS_Id           EventTimeStamp  \
0         1  990349  2015-02-21 10:47:13.000   
1         2  990360  2015-02-21 11:34:34.000   
2         3  990364  2015-02-21 11:35:31.000   
3         4  990370  2015-02-21 11:35:33.000   
4         5  990416  2015-02-21 11:39:41.000   

                          eventDescription  actionDescription  \
0  Low (Severity Low) Engine Coolant Level                NaN   
1                                      NaN                NaN   
2      Incorrect Data Steering Wheel Angle                NaN   
3      Incorrect Data Steering Wheel Angle                NaN   
4                                      NaN                NaN   

                     ecuSoftwareVersion ecuSerialNumber              ecuModel  \
0                               unknown         unknown               unknown   
1                               unknown         unknown               unknown   
2                               unknown         unknown               unknown   
3                               unknown         unknown               unknown   
4  22281684P01*22357957P01*22362082P01*        13063430  0USA13_13_0415_2238A   

   ecuMake  ecuSource   spn  fmi  active  activeTransitionCount  faultValue  \
0  unknown          0   111   17    True                      2         NaN   
1  unknown         11   629   12    True                    127         NaN   
2  unknown         11  1807    2   False                    127         NaN   
3  unknown         11  1807    2    True                    127         NaN   
4    VOLVO          0  4364   17   False                      2         NaN   

  EquipmentID  MCTNumber   Latitude  Longitude        LocationTimeStamp  
0        1439  105354361  38.857638 -84.626851  2015-02-21 11:34:25.000  
1        1439  105354361  38.857638 -84.626851  2015-02-21 11:35:10.000  
2        1369  105336226  41.421250 -87.767361  2015-02-21 11:35:26.000  
3        1369  105336226  41.421018 -87.767361  2015-02-21 11:36:08.000  
4        1674  105427130  38.416481 -89.442638  2015-02-21 11:39:37.000

RecordID     ESS_Id           EventTimeStamp  \
1187330   1248454  123904424  2020-03-06 14:00:26.000   
1187331   1248455  123905139  2020-03-06 14:04:23.000   
1187332   1248456  123905996  2020-03-06 14:13:38.000   
1187333   1248457  123906113  2020-03-06 14:14:13.000   
1187334   1248458  123906131  2020-03-06 14:15:34.000   

                                          eventDescription  actionDescription  \
1187330             Low (Severity Low) Catalyst Tank Level                NaN   
1187331   Condition Exists Engine Protection Torque Derate                NaN   
1187332  Abnormal Rate of Change Aftertreatment 1 Intak...                NaN   
1187333         Low (Severity Medium) Engine Coolant Level                NaN   
1187334         Low (Severity Medium) Engine Coolant Level                NaN   

                                      ecuSoftwareVersion ecuSerialNumber  \
1187330  04384413*22383729*082218154102*60701732*G1*BGT*        80156139   
1187331  04358814*06099720*030816202706*09400153*G1*BDR*        79932020   
1187332  05317106*05100987*050719120655*09401585*G1*BDR*        79880653   
1187333  04384413*22544852*090619141107*60701756*G1*BGT*             NaN   
1187334  04384413*22544852*090619141107*60701756*G1*BGT*             NaN   

                  ecuModel ecuMake  ecuSource   spn  fmi  active  \
1187330  6X1u17D1500000000   CMMNS          0  1761   17   False   
1187331  6X1u13D1500000000   CMMNS          0  1569   31    True   
1187332  6X1u13D1500000000   CMMNS          0  3216   10    True   
1187333                NaN     NaN          0   111   18    True   
1187334                NaN     NaN          0   111   18   False   

         activeTransitionCount  faultValue EquipmentID  MCTNumber   Latitude  \
1187330                      3         NaN        2282  105439740  37.094768   
1187331                      5         NaN        1994  105354084  34.390740   
1187332                      1         NaN        1850  105336308  34.430370   
1187333                      8         NaN        2377  108605700  35.030925   
1187334                      8         NaN        2377  108605700  35.027314   

         Longitude        LocationTimeStamp  
1187330 -85.897407  2020-03-06 14:00:21.000  
1187331 -79.461805  2020-03-06 14:04:59.000  
1187332 -84.920509  2020-03-06 14:14:14.000  
1187333 -85.321527  2020-03-06 14:14:49.000  
1187334 -85.323472  2020-03-06 14:15:30.000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187335 entries, 0 to 1187334
Data columns (total 20 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   RecordID               1187335 non-null  int64  
 1   ESS_Id                 1187335 non-null  int64  
 2   EventTimeStamp         1187335 non-null  object 
 3   eventDescription       1126490 non-null  object 
 4   actionDescription      0 non-null        float64
 5   ecuSoftwareVersion     891285 non-null   object 
 6   ecuSerialNumber        844318 non-null   object 
 7   ecuModel               1122577 non-null  object 
 8   ecuMake                1122577 non-null  object 
 9   ecuSource              1187335 non-null  int64  
 10  spn                    1187335 non-null  int64  
 11  fmi                    1187335 non-null  int64  
 12  active                 1187335 non-null  bool   
 13  activeTransitionCount  1187335 non-null  int64  
 14  faultValue        

None

### Provided Data Dictionary for `engine_faults`

Columns | Description
:-------|:-----------
`ESS_Id` | The event subscriber service event that contained the fault
`EventTimeStamp` | When the event took place
`eventDescription` | Brief text of meaning of the code (not always present)
`actionDescription` | Never seen this filled in
`ecuSoftwareVersion` | version string from the reporting vehicle computer system
`ecuSerialNumber` | Serial number of the reporting Engine Control Module (ECM)
`ecuModel` | Model of the reporting ECM
`ecuMake` | Manufacturer of the reporting ECM
`ecuSource` | 
`spn` | Fault code being reported
`fmi` | Failure Mode associated with the Fault Code
`active` | Whether the code is being set or being removed
`activeTransitionCount` | Number of times code has been set/unset
`faultValue` | Never seen used
`EquipmentID` | Assigned truck number of the unit in question
`MCTNumber` | Communications Terminal assigned to the truck
`Latitude` | Latitude at time of event
`Longitude` | Longitude at time of event
`LocationTimeStamp` | Time latitude and longitude were obtained

### Data Cleanups `engine_Faults`

In [7]:
# Drop these columns: There are no data in these columns (Just `NaN`)
engine_faults = engine_faults.drop(columns=["actionDescription", "faultValue"])

In [8]:
# Rename all columns to format "column_name"
engine_faults = engine_faults.rename(columns={
    "RecordID": "record_id",
    "ESS_Id": "ess_id",
    "EventTimeStamp": "event_timestamp",
    "eventDescription": "event_description",
    "actionDescription": "action_description",
    "ecuSoftwareVersion": "ecu_software_version",
    "ecuSerialNumber": "ecu_serial_number",
    "ecuModel": "ecu_model",
    "ecuMake": "ecu_make",
    "ecuSource": "ecu_source",
    "activeTransitionCount": "active_transition_count",
    "faultValue": "fault_value",
    "EquipmentID": "equipment_id",
    "MCTNumber": "mct_number",
    "Latitude": "latitude",
    "Longitude": "longitude",
    "LocationTimeStamp": "location_timestamp"
})

In [9]:
# Columns To String
to_string_columns = [
    "record_id", 
    "ess_id",
    "event_description",
    "ecu_software_version",
    "ecu_serial_number",
    "ecu_model",
    "ecu_make",
    "ecu_source",
    "spn",
    "fmi",
    "equipment_id",
    "mct_number",
    #"latitude",
    #"longitude"
]

for col in to_string_columns:
    engine_faults[col] = engine_faults[col].astype("str")

In [10]:
# Columns To Boolean
engine_faults["active"] = engine_faults["active"].astype("bool")

In [11]:
# Columns To Integer
engine_faults["active_transition_count"] = engine_faults["active_transition_count"].astype("int64")

In [12]:
# Columns To Float
engine_faults["latitude"] = engine_faults["latitude"].astype("float64")
engine_faults["longitude"] = engine_faults["longitude"].astype("float64")

In [13]:
# Columns To datetime
engine_faults["event_timestamp"] = engine_faults["event_timestamp"].astype("datetime64")
engine_faults["location_timestamp"] = engine_faults["location_timestamp"].astype("datetime64")

In [14]:
# Check where we are
display(engine_faults.shape)

(1187335, 18)

#### Remove faults where equipmentID has more than 5 characters

In [15]:
engine_faults = engine_faults[engine_faults["equipment_id"].map(len) <= 5]

In [16]:
# Check where we are
display(engine_faults.shape)

(1185166, 18)

#### Remove faults in the vicinity of service location

The service locations:

- `(36.0666667, -86.4347222)`
- `(35.5883333, -86.4438888)`
- `(36.1950, -83.174722)`

**Coordinates Precision**

Decimal Places | Decimal Degrees | Object that can be unambiguously recognized at this scale
:-------------:|-----------------|:---------------------------------------------------------
0 | 1.0 | country or large region
1 | 0.1 | large city or district
2 | 0.01 | town or village
3 | 0.001 | neighborhood, street
4 | 0.0001 | individual street, large buildings
5 | 0.00001 | individual trees, houses
6 | 0.000001 | individual humans
7 | 0.0000001 | practical limit of commercial surveying
8 | 0.00000001 | specialized surveying (e.g. tectonic plate mapping)

In [17]:
# Work to be done here

# Solution 1: Use Decimal Degrees Precision of Coordinates

# Solution 2: Use Haversine method (Use numpy)
# https://www.geeksforgeeks.org/program-distance-two-points-earth/

#### Remove faults where `active` is set to `False`

In [18]:
engine_faults = engine_faults[engine_faults["active"] == False]

In [19]:
# Check where we are
display(engine_faults.shape)

(577830, 18)

### Final `engine_faults` before merge

In [20]:
display(engine_faults.shape)
display(engine_faults.head())
display(engine_faults.tail())
display(engine_faults.info())

(577830, 18)

record_id  ess_id     event_timestamp  \
2          3  990364 2015-02-21 11:35:31   
4          5  990416 2015-02-21 11:39:41   
9         10  990446 2015-02-21 11:41:55   
11        12  990462 2015-02-21 11:43:18   
12        13  990468 2015-02-21 11:43:10   

                          event_description  \
2       Incorrect Data Steering Wheel Angle   
4                                       nan   
9   Low (Severity Low) Engine Coolant Level   
11  Low (Severity Low) Engine Coolant Level   
12  Low (Severity Low) Engine Coolant Level   

                               ecu_software_version ecu_serial_number  \
2                                           unknown           unknown   
4              22281684P01*22357957P01*22362082P01*          13063430   
9   04993120*00025921*082113134117*07700053*I0*BBZ*          79466580   
11                                          unknown           unknown   
12                                          unknown           unknown   

               ecu_model ecu_make ecu_source   spn fmi  active  \
2                unknown  unknown         11  1807   2   False   
4   0USA13_13_0415_2238A    VOLVO          0  4364  17   False   
9      6X1u10D1500000000    CMMNS          0   111  17   False   
11               unknown  unknown          0   111  17   False   
12               unknown  unknown          0   111  17   False   

    active_transition_count equipment_id mct_number   latitude  longitude  \
2                       127         1369  105336226  41.421250 -87.767361   
4                         2         1674  105427130  38.416481 -89.442638   
9                         1         1417  105438630  33.039953 -96.182592   
11                        2         1439  105354361  38.857592 -84.626805   
12                        2         1597  105344243  36.881527 -86.444722   

    location_timestamp  
2  2015-02-21 11:35:26  
4  2015-02-21 11:39:37  
9  2015-02-21 11:41:51  
11 2015-02-21 11:43:13  
12 2015-02-21 11:43:06

record_id     ess_id     event_timestamp  \
1187326   1248450  123900153 2020-03-06 13:33:41   
1187327   1248451  123900154 2020-03-06 13:33:41   
1187329   1248453  123901925 2020-03-06 13:44:45   
1187330   1248454  123904424 2020-03-06 14:00:26   
1187334   1248458  123906131 2020-03-06 14:15:34   

                                  event_description  \
1187326                   High Voltage (Fuel Level)   
1187327       High Voltage (Left Fuel Level Sensor)   
1187329  Low (Severity Medium) Engine Coolant Level   
1187330      Low (Severity Low) Catalyst Tank Level   
1187334  Low (Severity Medium) Engine Coolant Level   

                                    ecu_software_version ecu_serial_number  \
1187326                                              nan               nan   
1187327                                              nan               nan   
1187329  04358814*06030918*051718174436*09401683*G1*BDR*          79904453   
1187330  04384413*22383729*082218154102*60701732*G1*BGT*          80156139   
1187334  04384413*22544852*090619141107*60701756*G1*BGT*               nan   

                 ecu_model ecu_make ecu_source   spn fmi  active  \
1187326      CECU3B-NAMUX4    PACCR         49    96   3   False   
1187327      CECU3B-NAMUX4    PACCR         49   829   3   False   
1187329  6X1u13D1500000000    CMMNS          0   111  18   False   
1187330  6X1u17D1500000000    CMMNS          0  1761  17   False   
1187334                nan      nan          0   111  18   False   

         active_transition_count equipment_id mct_number   latitude  \
1187326                      126         1936  105355619  30.376898   
1187327                      126         1936  105355619  30.376898   
1187329                       93         1886  105351219  39.013472   
1187330                        3         2282  105439740  37.094768   
1187334                        8         2377  108605700  35.027314   

         longitude  location_timestamp  
1187326 -81.744814 2020-03-06 13:33:36  
1187327 -81.744814 2020-03-06 13:33:36  
1187329 -77.037685 2020-03-06 13:44:41  
1187330 -85.897407 2020-03-06 14:00:21  
1187334 -85.323472 2020-03-06 14:15:30

<class 'pandas.core.frame.DataFrame'>
Int64Index: 577830 entries, 2 to 1187334
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   record_id                577830 non-null  object        
 1   ess_id                   577830 non-null  object        
 2   event_timestamp          577830 non-null  datetime64[ns]
 3   event_description        577830 non-null  object        
 4   ecu_software_version     577830 non-null  object        
 5   ecu_serial_number        577830 non-null  object        
 6   ecu_model                577830 non-null  object        
 7   ecu_make                 577830 non-null  object        
 8   ecu_source               577830 non-null  object        
 9   spn                      577830 non-null  object        
 10  fmi                      577830 non-null  object        
 11  active                   577830 non-null  bool          
 12  active_transiti

None

## Looking at `diagnostics`

In [21]:
display(diagnostics.shape)
display(diagnostics.head(5))
display(diagnostics.tail(5))
display(diagnostics.info())

(12821626, 4)

Id                  Name     Value  FaultId
0   1             IgnStatus     False        1
1   2     EngineOilPressure         0        1
2   3  EngineOilTemperature  96.74375        1
3   4    TurboBoostPressure         0        1
4   5            EngineLoad        11        1

Id                      Name        Value  FaultId
12821621  12864020  EngineCoolantTemperature        181.4  1248457
12821622  12864021              ParkingBrake        False  1248457
12821623  12864022    SwitchedBatteryVoltage         14.1  1248457
12821624  12864023               DistanceLtd  28606.65625  1248457
12821625  12864024                LampStatus         1023  1248458

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12821626 entries, 0 to 12821625
Data columns (total 4 columns):
 #   Column   Dtype 
---  ------   ----- 
 0   Id       int64 
 1   Name     object
 2   Value    object
 3   FaultId  int64 
dtypes: int64(2), object(2)
memory usage: 391.3+ MB


None

### Data Cleanups: `diagnostics`

In [22]:
# Pivotting
diagnostics = diagnostics.pivot(index = 'FaultId', columns='Name', values='Value').reset_index()

In [23]:
# Rename all columns to format "column_name"
diagnostics = diagnostics.rename(columns={
    "FaultId": "fault_id",
    "AcceleratorPedal": "accelerator_pedal",
    "BarometricPressure": "barometric_pressure",
    "CruiseControlActive": "cruise_control_active",
    "CruiseControlSetSpeed": "cruise_control_set_speed",
    "DistanceLtd": "distance_ltd",
    "EngineCoolantTemperature": "engine_coolant_temperature",
    "EngineLoad": "engine_load",
    "EngineOilPressure": "engine_oil_pressure",
    "EngineOilTemperature": "engine_oil_temperature",
    "EngineRpm": "engine_rpm",
    "EngineTimeLtd": "engine_time_ltd",
    "FuelLevel": "fuel_level",
    "FuelLtd": "fuel_ltd",
    "FuelRate": "fuel_rate",
    "FuelTemperature": "fuel_temperature",
    "IgnStatus": "ign_status",
    "IntakeManifoldTemperature": "intake_manifold_temperature",
    "LampStatus": "lamp_status",
    "ParkingBrake": "parking_brake",
    "ServiceDistance": "service_distance",
    "Speed": "speed",
    "SwitchedBatteryVoltage": "switched_battery_voltage",
    "Throttle": "throttle",
    "TurboBoostPressure": "turbo_boost_pressure"
})

In [24]:
# Columns To String
diagnostics["fault_id"] = diagnostics["fault_id"].astype("str")
diagnostics["lamp_status"] = diagnostics["lamp_status"].astype("str")

In [25]:
# Columns To Boolean
diagnostics["cruise_control_active"] = diagnostics["cruise_control_active"].astype("bool")
diagnostics["ign_status"] = diagnostics["ign_status"].astype("bool")
diagnostics["parking_brake"] = diagnostics["parking_brake"].astype("bool")

In [26]:
# Columns To Float
to_float_columns = [
    "accelerator_pedal", 
    "barometric_pressure",
    "cruise_control_set_speed",
    "distance_ltd",
    "engine_coolant_temperature",
    "engine_load",
    "engine_oil_pressure",
    "engine_oil_temperature",
    "engine_rpm",
    "engine_time_ltd",
    "fuel_level",
    "fuel_ltd",
    "fuel_rate",
    "fuel_temperature",
    "intake_manifold_temperature",
    "service_distance",
    "speed",
    "switched_battery_voltage",
    "throttle",
    "turbo_boost_pressure"
]

for col in to_float_columns:
    diagnostics[col] = diagnostics[col].str.replace(',', '.')
    diagnostics[col] = diagnostics[col].astype("float64")

In [27]:
# Check where we are
display(diagnostics.shape)
display(diagnostics.head())
display(diagnostics.tail())
display(diagnostics.info())

(1187335, 25)

Name fault_id  accelerator_pedal  barometric_pressure  cruise_control_active  \
0           1                0.0                14.21                   True   
1           2                NaN                  NaN                   True   
2           3                NaN                  NaN                   True   
3           4                NaN                  NaN                   True   
4           5                NaN                  NaN                   True   

Name  cruise_control_set_speed  distance_ltd  engine_coolant_temperature  \
0                     66.48672      423178.7                       100.4   
1                          NaN           NaN                         NaN   
2                          NaN           NaN                         NaN   
3                          NaN           NaN                         NaN   
4                          NaN           NaN                         NaN   

Name  engine_load  engine_oil_pressure  engine_oil_temperature  ...  \
0            11.0                  0.0                96.74375  ...   
1             NaN                  NaN                     NaN  ...   
2             NaN                  NaN                     NaN  ...   
3             NaN                  NaN                     NaN  ...   
4             NaN                  NaN                     NaN  ...   

Name  fuel_temperature  ign_status  intake_manifold_temperature  lamp_status  \
0                  NaN        True                         78.8         1023   
1                  NaN        True                          NaN         1279   
2                  NaN        True                          NaN         1279   
3                  NaN        True                          NaN         1279   
4                  NaN        True                          NaN        16639   

Name  parking_brake  service_distance  speed  switched_battery_voltage  \
0              True               NaN    0.0                   3276.75   
1              True               NaN    NaN                       NaN   
2              True               NaN    NaN                       NaN   
3              True               NaN    NaN                       NaN   
4              True               NaN    NaN                       NaN   

Name throttle  turbo_boost_pressure  
0         NaN                   0.0  
1         NaN                   NaN  
2         NaN                   NaN  
3         NaN                   NaN  
4         NaN                   NaN  

[5 rows x 25 columns]

Name    fault_id  accelerator_pedal  barometric_pressure  \
1187330  1248454                NaN                  NaN   
1187331  1248455              100.0              14.5000   
1187332  1248456                0.0              14.3550   
1187333  1248457                1.6              14.4275   
1187334  1248458                NaN                  NaN   

Name     cruise_control_active  cruise_control_set_speed  distance_ltd  \
1187330                   True                       NaN           NaN   
1187331                   True                  64.62260  423937.90000   
1187332                   True                  66.48672  465925.40000   
1187333                   True                  67.72946   28606.65625   
1187334                   True                       NaN           NaN   

Name     engine_coolant_temperature  engine_load  engine_oil_pressure  \
1187330                         NaN          NaN                  NaN   
1187331                       185.0         51.0                37.12   
1187332                       186.8         62.0                41.18   
1187333                       181.4          0.0                27.26   
1187334                         NaN          NaN                  NaN   

Name     engine_oil_temperature  ...  fuel_temperature  ign_status  \
1187330                     NaN  ...               NaN        True   
1187331                211.4937  ...              32.0        True   
1187332                212.8438  ...               NaN        True   
1187333                221.7312  ...               NaN        True   
1187334                     NaN  ...               NaN        True   

Name     intake_manifold_temperature  lamp_status  parking_brake  \
1187330                          NaN         1023           True   
1187331                         98.6        18431           True   
1187332                         91.4        17407           True   
1187333                        100.4         1023           True   
1187334                          NaN         1023           True   

Name     service_distance     speed  switched_battery_voltage throttle  \
1187330               NaN       NaN                       NaN      NaN   
1187331               NaN  65.01096                       NaN     73.2   
1187332               NaN  66.57410                       NaN    100.0   
1187333               NaN  11.84489                      14.1    100.0   
1187334               NaN       NaN                       NaN      NaN   

Name     turbo_boost_pressure  
1187330                   NaN  
1187331                  7.83  
1187332                  6.96  
1187333                  1.74  
1187334                   NaN  

[5 rows x 25 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187335 entries, 0 to 1187334
Data columns (total 25 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   fault_id                     1187335 non-null  object 
 1   accelerator_pedal            531889 non-null   float64
 2   barometric_pressure          585976 non-null   float64
 3   cruise_control_active        1187335 non-null  bool   
 4   cruise_control_set_speed     576458 non-null   float64
 5   distance_ltd                 585819 non-null   float64
 6   engine_coolant_temperature   586071 non-null   float64
 7   engine_load                  585621 non-null   float64
 8   engine_oil_pressure          586244 non-null   float64
 9   engine_oil_temperature       583912 non-null   float64
 10  engine_rpm                   586921 non-null   float64
 11  engine_time_ltd              581366 non-null   float64
 12  fuel_level                   502795 non-nu

None

In [ ]:
# Haversine
# Python 3 program to calculate Distance Between Two Points on Earth
from math import radians, cos, sin, asin, sqrt
def distance(lat1, lat2, lon1, lon2):
     
    # The math module contains a function named
    # radians which converts from degrees to radians.
    lon1 = radians(lon1)
    lon2 = radians(lon2)
    lat1 = radians(lat1)
    lat2 = radians(lat2)
      
    # Haversine formula
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
 
    c = 2 * asin(sqrt(a))
    
    # Radius of earth in kilometers. Use 3956 for miles
    r = 6371
      
    # calculate the result
    return(c * r)
     
     
# driver code
lat1 = 53.32055555555556
lat2 = 53.31861111111111
lon1 = -1.7297222222222221
lon2 =  -1.6997222222222223
print(distance(lat1, lat2, lon1, lon2), "K.M")

## Looking At `service_fault_codes`

In [28]:
display(service_fault_codes.shape)
display(service_fault_codes.head(5))
display(service_fault_codes.info())

(7124, 14)

Published in CES 14602  Cummins Fault Code  Revision         PID  \
0                      Y                 111       167  Not Mapped   
1                      Y                 112       167  Not Mapped   
2                      Y                 113       167  Not Mapped   
3                      Y                 114       167  Not Mapped   
4                      Y                 115       167         190   

          SID         MID  J1587 FMI  SPN  J1939 FMI J2012 Pcode Lamp Color  \
0         254           0         12  629         12       P0606        Red   
1          20         128          7  635          7  Not Mapped        Red   
2          20         128          3  635          3  Not Mapped      Amber   
3          20         128          4  635          4  Not Mapped      Amber   
4  Not Mapped  Not Mapped          2  612          2       P0008        Red   

       Lamp Device                                Cummins Description  \
0  Stop / Shutdown  Engine Control Module Critical Internal Failur...   
1  Stop / Shutdown  Engine Timing Actuator Driver Circuit - Mechan...   
2          Warning  Engine Timing Actuator Driver Circuit - Voltag...   
3          Warning  Engine Timing Actuator Driver Circuit - Voltag...   
4  Stop / Shutdown  Engine Magnetic Speed/Position Lost Both of Tw...   

                               Algorithm Description  
0  Error internal to the ECM related to memory ha...  
1  Mechanical failure in the engine timing actuat...  
2  High signal voltage detected at the engine tim...  
3  Low voltage detected at the engine timing actu...  
4  The ECM has detected that the primary and back...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7124 entries, 0 to 7123
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Published in CES 14602  7124 non-null   object
 1   Cummins Fault Code      7124 non-null   int64 
 2   Revision                7124 non-null   int64 
 3   PID                     7124 non-null   object
 4   SID                     7124 non-null   object
 5   MID                     7124 non-null   object
 6   J1587 FMI               7124 non-null   int64 
 7   SPN                     7124 non-null   int64 
 8   J1939 FMI               7124 non-null   int64 
 9   J2012 Pcode             7124 non-null   object
 10  Lamp Color              7124 non-null   object
 11  Lamp Device             7124 non-null   object
 12  Cummins Description     7124 non-null   object
 13  Algorithm Description   2005 non-null   object
dtypes: int64(5), object(9)
memory usage: 779.3+ KB


None

### Data Cleanups `service_fault_codes`

In [29]:
# Rename all columns to format "column_name"
service_fault_codes = service_fault_codes.rename(columns={
    "Published in CES 14602": "published_in_ces14602",
    "Cummins Fault Code": "cummins_fault_code",
    "Revision": "revision",
    "PID": "p_id",
    "SID": "s_id",
    "MID": "m_id",
    "J1587 FMI": "j1587_fmi",
    "SPN": "spn",
    "J1939 FMI": "j1939_fmi",
    "J2012 Pcode": "j2012_pcode",
    "Lamp Color": "lamp_color",
    "Lamp Device": "lamp_device",
    "Cummins Description": "cummins_description",
    "Algorithm Description": "algorithm_description"
})

- `published_in_ces14602`: `Y` to all 7124
- `cummins_fault_code`: Each one is a unique fault code
- `revision`: Only one revision : 167
- `p_id`: Most are `Not Mapped`
- `s_id`: Most are `Not Mapped`
- `m_id`: Most are `Not Mapped`, `0`, or `128`
- `j1587_fmi`: From `0` to `14`, most are 11
- `spn`: 1915 rows
- `j1939_fmi`: From `0` to `31`, most are 31
- `j2012_pcode`: Most are not mapped
- `lamp_color`: Most are `Amber`, but also `Maintenance`, `None`, `Red`
- `lamp_device`: Most are `Warning`, but also `Maintenance`, `None`, `Stop / Shutdown`
- `cummins_description`: 5374 unique entries
- `algorithm_description`: 1879 unique entries